module/zfs/brt.c

```diff
@@ -0,0 +1,1817 @@
+{
+	brt_vdev_t *brtvd;
+	uint64_t mos_entries = 0;
+
+	brt_rlock(brt);
+	brtvd = brt_vdev(brt, vdevid);
+	if (brtvd != NULL)
+		mos_entries = brtvd->bv_mos_entries;
+	brt_unlock(brt);
+
+	if (mos_entries == 0)
+		return;
+
+	BRT_DEBUG("ZAP prefetch: object=%llu vdev=%llu offset=%llu",
+	    mos_entries, vdevid, bre->bre_offset);
+	(void) zap_prefetch_uint64(brt->brt_mos, mos_entries,
+	    (uint64_t *)&bre->bre_offset, BRT_KEY_WORDS);
+}
+
+static int
+brt_entry_update(brt_t *brt, brt_vdev_t *brtvd, brt_entry_t *bre, dmu_tx_t *tx)
+{
+	int error;
+
+	ASSERT(RW_LOCK_HELD(&brt->brt_lock));
+	ASSERT(brtvd->bv_mos_entries != 0);
+	ASSERT(bre->bre_refcount > 0);
+
+	error = zap_update_uint64(brt->brt_mos, brtvd->bv_mos_entries,
+	    (uint64_t *)&bre->bre_offset, BRT_KEY_WORDS, 1,
+	    sizeof (bre->bre_refcount), &bre->bre_refcount, tx);
+	BRT_DEBUG("ZAP update: object=%llu vdev=%llu offset=%llu count=%llu "
+	    "error=%d", brtvd->bv_mos_entries, brtvd->bv_vdevid,
+	    bre->bre_offset, bre->bre_refcount, error);
+
+	return (error);
+}
+
+static int
+brt_entry_remove(brt_t *brt, brt_vdev_t *brtvd, brt_entry_t *bre, dmu_tx_t *tx)
+{
+	int error;
+
+	ASSERT(RW_LOCK_HELD(&brt->brt_lock));
+	ASSERT(brtvd->bv_mos_entries != 0);
+	ASSERT0(bre->bre_refcount);
+
+	error = zap_remove_uint64(brt->brt_mos, brtvd->bv_mos_entries,
+	    (uint64_t *)&bre->bre_offset, BRT_KEY_WORDS, tx);
+	BRT_DEBUG("ZAP remove: object=%llu vdev=%llu offset=%llu count=%llu "
+	    "error=%d", brtvd->bv_mos_entries, brtvd->bv_vdevid,
+	    bre->bre_offset, bre->bre_refcount, error);
+
+	return (error);
+}
+
+/*
+ * Return TRUE if we _can_ have BRT entry for this bp. It might be false
+ * positive, but gives us quick answer if we should look into BRT, which
+ * may require reads and thus will be more expensive.
+ */
+boolean_t
+brt_maybe_exists(spa_t *spa, const blkptr_t *bp)
+{
+	brt_t *brt = spa->spa_brt;
+	brt_vdev_t *brtvd;
+	brt_entry_t bre_search;
+	boolean_t mayexists = FALSE;
+	uint64_t vdevid;
+
+	brt_entry_fill(bp, &bre_search, &vdevid);
+
+	brt_rlock(brt);
+
+	brtvd = brt_vdev(brt, vdevid);
+	if (brtvd != NULL && brtvd->bv_initiated) {
+		if (!avl_is_empty(&brtvd->bv_tree) ||
+		    brt_vdev_lookup(brt, brtvd, &bre_search)) {
+			mayexists = TRUE;
+		}
+	}
+
+	brt_unlock(brt);
+
+	return (mayexists);
+}
+
+uint64_t
+brt_get_dspace(spa_t *spa)
+{
+	brt_t *brt = spa->spa_brt;
+
+	if (brt == NULL)
+		return (0);
+
+	return (brt->brt_dsize);
+}
+
+uint64_t
+brt_get_used(spa_t *spa)
+{
+	brt_t *brt = spa->spa_brt;
+
+	if (brt == NULL)
+		return (0);
+
+	return (brt->brt_drefsize);
+}
+
+uint64_t
+brt_get_logical_used(spa_t *spa)
+{
+	brt_t *brt = spa->spa_brt;
+
+	if (brt == NULL)
+		return (0);
+
+	return (brt->brt_drefsize + brt->brt_dsize);
+}
+
+uint64_t
+brt_get_ratio(spa_t *spa)
+{
+	brt_t *brt = spa->spa_brt;
+
+	if (brt->brt_dsize == 0)
+		return (100);
+
+	return ((brt->brt_drefsize + brt->brt_dsize) * 100 / brt->brt_drefsize);
```

Comment:
Possible division by zero if `brt->brt_drefsize` isn't checked.

Reply:
It is not possible for brt_drefsize to be 0 if brt_dsize != 0, but to make it more obvious I've changed the check above to test brt_drefsize.